zebROS_ws/src/behaviors/src/2023_align_to_grid.cpp

```diff
@@ -0,0 +1,280 @@
+// so we can do the equivalent by calculating the offset between a detected apriltag (see 2023_apriltag_locations.yaml) and the desired location
+// and averaging if there are multiple.
+// find closest apriltag to desired location, calculate offset, call path_to_apriltag
+#include "ros/ros.h"
+#include "field_obj/Detection.h"
+#include "geometry_msgs/Point.h"
+#include <ros/ros.h>
+#include <actionlib/server/simple_action_server.h>
+#include "behavior_actions/AlignToGrid2023Action.h"
+#include "behavior_actions/PathToAprilTagAction.h"
+#include <actionlib/client/simple_action_client.h>
+#include <tf2/LinearMath/Quaternion.h>
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+
+inline geometry_msgs::Point operator-(const geometry_msgs::Point& lhs, const geometry_msgs::Point& rhs) {
+    geometry_msgs::Point p;
+    p.x = lhs.x - rhs.x;
+    p.y = lhs.y - rhs.y;
+    p.z = lhs.z - rhs.z;
+    return p;
+}
+
+inline geometry_msgs::Point operator+(const geometry_msgs::Point& lhs, const geometry_msgs::Point& rhs) {
+    geometry_msgs::Point p;
+    p.x = lhs.x + rhs.x;
+    p.y = lhs.y + rhs.y;
+    p.z = lhs.z + rhs.z;
+    return p;
+}
+
+double distance(const geometry_msgs::Point &pt) {
+    return sqrt(pt.x * pt.x + pt.y * pt.y + pt.z * pt.z);
+}
+
+inline bool operator<(const geometry_msgs::Point& lhs, const geometry_msgs::Point& rhs) {
+    return distance(lhs) < distance(rhs);
+}
+
+inline bool operator>(const geometry_msgs::Point& lhs, const geometry_msgs::Point& rhs) {
+    return distance(lhs) > distance(rhs);
+}
+
+class Tag {
+public:
+	geometry_msgs::Point location;
+	double rotation;
+	Tag(geometry_msgs::Point location = geometry_msgs::Point(), bool _rotation = 0) {
+		this->location = location;
+		this->rotation = _rotation;
+	}
+};
+
+enum GamePiece { CUBE, CONE };
+
+struct GridLocation {
+  double y;
+  GamePiece piece;
+  GridLocation(double y = 0, GamePiece piece = CUBE) {
+    this->y = y;
+    this->piece = piece;
+  }
+};
+
+class AlignToGridAction
+{
+protected:
+
+  ros::NodeHandle nh_;
+  actionlib::SimpleActionServer<behavior_actions::AlignToGrid2023Action> as_; // NodeHandle instance must be created before this line. Otherwise strange error occurs.
+  std::string action_name_;
+  // create messages that are used to published feedback/result
+  behavior_actions::AlignToGrid2023Feedback feedback_;
+  behavior_actions::AlignToGrid2023Result result_;
+  field_obj::Detection latest_;
+  ros::Subscriber sub_;
+  std::map<int, Tag> tags_;
+  std::map<int, GridLocation> gridLocations_; // should probably be uint8_t
+  actionlib::SimpleActionClient<behavior_actions::PathToAprilTagAction> client_;
+
+public:
+
+  AlignToGridAction(std::string name) :
+    as_(nh_, name, boost::bind(&AlignToGridAction::executeCB, this, _1), false),
+    action_name_(name),
+    sub_(nh_.subscribe<field_obj::Detection>("/tf_object_detection/tag_detection_world", 1000, boost::bind(&AlignToGridAction::callback, this, _1))),
```

Comment:
Set the queue size to 1 - no need to process previous detections, just the most current

Reply:
cool, thanks! just changed it